pkg/review/commit_review_status.go

```diff
@@ -192,12 +207,59 @@ func (fn *CommitApprovalDoFn) ProcessElement(ctx context.Context, commit Commit,
 	emit(commitReviewStatus)
 }
 
+// ProcessElement is a DoFn implementation that takes a CommitReviewStatus
+// and populates its breakGlassIssue field (if necessary) and then emits it
+// using the given emit function. The process only searches for break glass
+// issues for commits that do not have the status GithubPRApproved.
+func (fn *BreakGlassIssueDoFn) ProcessElement(ctx context.Context, commitReviewStatus CommitReviewStatus, emit func(status CommitReviewStatus)) {
+	logger := logging.FromContext(ctx)
+	logger.InfoContext(ctx, "processing commitReviewStatus", "commitReviewStatus", commitReviewStatus)
+	if commitReviewStatus.ApprovalStatus != GithubPRApproved {
+		// if the commit does not have proper approval, we check if there was a
+		// break glass issue opened by the author during the timeframe they
+		// submitted the commit.
+		breakGlassIssue, err := getBreakGlassIssue(
+			ctx,
+			fn.Client,
+			fn.Config.IssuesTable,
+			commitReviewStatus.Author,
+			commitReviewStatus.Timestamp,
+		)
+		if err != nil {
+			// We should only get transient style errors from BigQuery
+			// (e.g. network is down etc.). So, we can just log the error and then
+			// drop this CommitReviewStatus from the pipeline. It will then get
+			// retried on the next run of the pipeline.
+			logger.ErrorContext(ctx, "failure when trying to get break glass issue: %v", err)
+			return
+		}
+		commitReviewStatus.BreakGlassURL = breakGlassIssue.HTMLURL
+	}
+	emit(commitReviewStatus)
+}
+
+// getBreakGlassIssue Searches BigQuery for a break glass issue that was opened
+// by the given author and whose created_at date was on or before the given
+// timestamp and whose closed_at date was on or after the given timestamp.
+func getBreakGlassIssue(ctx context.Context, client bq.Client[breakGlassIssue], issuesTable bigqueryio.QualifiedTableName, author, timestamp string) (*breakGlassIssue, error) {
```

Comment:
Why wouldn't this just be a function on our client interface instead?

Reply:
Because the client is generic and this function is is mostly related to my specific business logic of getting the break glass issue. I could refactor the parts pertaining to executing a query that you only expect to have 1 into the client. Something like:
```
func (bqc *bigQueryClientImpl[T]) QueryAndGetFirst(ctx context.Context, queryString string) (*T, error) {
	items, err := bqc.Query(ctx, queryString)
	if err != nil {
		return nil, fmt.Errorf("bqc.Query failed: %w", err)
	}
	var t *T
	if len(items) > 0 {
		t = &items[0]
	}
	return t, nil
}
```
Then this function could be rewritten as:
```
func getBreakGlassIssue(ctx context.Context, client bq.Client[breakGlassIssue], issuesTable bigqueryio.QualifiedTableName, author, timestamp string) (*breakGlassIssue, error) {
	issueQuery := GetBreakGlassIssueQuery(issuesTable, author, timestamp)
	breakGlassIssue, err := client.QueryAndGetFirst(ctx, issueQuery)
	if err != nil {
		return nil, fmt.Errorf("client.QueryAndGetFirst failed: %w", err)
	}
	return breakGlassIssue, nil
}
```
I'm not sure if this is significantly better or not, but you can let me know what you think.

Reply:
Similar to what [I said here](https://github.com/abcxyz/github-metrics-aggregator/pull/151#discussion_r1361413242), I think we're conflating a generic client with an interface that abstracts business logic. I think you want your interface to be at a _higher_ level than anything BigQuery or database-specific.

Reply:
I've update the interface to be at a higher level of abstraction. Let me know what you think